### PR TITLE
feat(frontent): Expose new Telemetry data

### LINKED ISSUE #128: Expose new Telemetry data
## Story

As a chain developer who uses Nodes page, I want to see detailed information about nodes running the network so I can get a better idea about what is going on.

Nearcore is going to update the Telemetry format (https://github.com/nearprotocol/nearcore/pull/1751), so it is going to be:

```json
{
  "agent": {
    "name": "near-rs",
    "version": "0.4.5",
    "build": "6a527109-modified"
  },
  "chain": {
    "node_id": "ed25519:67U1Mup5E98t5fQTJvpMuS7bq8rYDmyqWjeQ5884LHH7",
    "status": "NoSync",
    "account_id": "test.near",
    "num_peers": 0,
    "is_validator": true,
    "latest_block_hash": "7a4d3SNNb1XGxqhHAKj5nGx8C1R4X8pdt8jobC9z3nup",
    "latest_block_height": 1889
  },
  "signature": "ed25519:4n5VzXijHUvgwroD1W6BTysRGKTMdPv6ks32kEUNGyb9recDYRpt2L8Y2Uq5ABPt4LZHFBAv2DPj1CAPGkwqkZg2",
  "system": {
    "bandwidth_download": 0,
    "bandwidth_upload": 0,
    "cpu_usage": 0.6015791296958923,
    "memory_usage": 42784
  }
}
```

So, the data is going to be split into three sections now (agent [info about the running binary], system [info about the system load], chain), and the only new data here is the "agent" section.

## Acceptance Criteria

- [x] Accept the new Telemetry format while keeping compatibility with the old one for a little while
- [ ] Display the new information (`chain.is_validator`, `agent.*`) in the UI
- [ ] Mobile ok
- [ ] Tested

## Designs

https://www.figma.com/file/jS1bjIiVxoEPNOYeXwddjH/NEAR-Explorer?node-id=186%3A15

## Notes

Nearcore sends Telemetry data as JSON with `POST` request to `/api/nodes`, which is [handled by Next.js and passed with an extra `ip_address` information to the backend](https://github.com/nearprotocol/near-explorer/blob/5ab5c54fe0e98a8fa4128720cc5537754526a3ca/frontend/src/pages/api/nodes.js), where [it is saved to the database](https://github.com/nearprotocol/near-explorer/blob/5ab5c54fe0e98a8fa4128720cc5537754526a3ca/backend/src/wamp.js#L14-L24). A lot of information is currently ignored, so we should add more columns or JSON column to store all the information.

The current ("old") Telemetry format is:

```json
{
  "node_id": "ed25519:67U1Mup5E98t5fQTJvpMuS7bq8rYDmyqWjeQ5884LHH7",
  "status": "NoSync",
  "account_id": "test.near",
  "num_peers": 0,
  "is_validator": true,
  "latest_block_hash": "7a4d3SNNb1XGxqhHAKj5nGx8C1R4X8pdt8jobC9z3nup",
  "latest_block_height": 1889,
  "signature": "ed25519:4n5VzXijHUvgwroD1W6BTysRGKTMdPv6ks32kEUNGyb9recDYRpt2L8Y2Uq5ABPt4LZHFBAv2DPj1CAPGkwqkZg2",
  "bandwidth_download": 0,
  "bandwidth_upload": 0,
  "cpu_usage": 0.6015791296958923,
  "memory": 42784
}
```

### PR DESCRIPTION
**Pull request by icerove**
fix #128

### CODE REVIEW

File: frontend/src/components/nodes/__test__/__snapshots__/NodeRow.test.tsx.snap [hunk 1]
```diff
@@ -39,36 +47,39 @@ exports[`<NodeRow /> renders 1`] = `
...
+              className="jsx-990356129"
+              src="/static/images/icon-m-size.svg"
               style={
                 Object {
                   "width": "12px",
                 }
               }
             />
-             @
+            @  null | ver.  null build  null
```
Comment (frol) [inline frontend/src/components/nodes/__test__/__snapshots__/NodeRow.test.tsx.snap:61]: Let's update the test to pass all the relevant data

File: frontend/src/libraries/explorer-wamp/nodes.ts [hunk 2]
```diff
@@ -23,10 +24,10 @@ export default class NodesApi extends ExplorerApi {
         `SELECT ip_address as ipAddress, moniker, account_id as accountId, node_id as nodeId, signature, 
                 last_seen as lastSeen, last_height as lastHeight, last_hash as lastHash,
                 agent_name as agentName, agent_version as agentVersion, agent_build as agentBuild,
-                peer_count as peerCount, is_validator as isValidator
+                peer_count as peerCount, is_validator as isValidator, status
                     FROM nodes
                     ${endTimestamp ? `WHERE last_seen < :endTimestamp` : ""}
-                    ORDER BY last_seen DESC
+                    ORDER BY is_validator, last_seen DESC
```
Comment (frol) [inline frontend/src/libraries/explorer-wamp/nodes.ts:30]: I don't think we need this as this will bring old validator nodes up.

File: frontend/src/libraries/explorer-wamp/nodes.ts [hunk 3]
```diff
@@ -39,4 +40,18 @@ export default class NodesApi extends ExplorerApi {
       throw error;
     }
   }
+
+  async getTotalValidators() {
+    try {
+      return await this.call("select", [
+        `SELECT COUNT(*) as total from nodes
```
Comment (frol) [inline frontend/src/libraries/explorer-wamp/nodes.ts:47]: Can we always have `FROM` spelled in upper case?

File: frontend/src/libraries/explorer-wamp/nodes.ts (hunk 2 above)
Comment (icerove) [reply frontend/src/libraries/explorer-wamp/nodes.ts:30]: I change the sequence since we still want sort by isValidator.
`ORDER BY last_seen, is_validator DESC`

Comment (frol) [reply frontend/src/libraries/explorer-wamp/nodes.ts]: Based on the Illia's request, we should have just two tabs: `WHERE is_validator = 1` and `WHERE is_validator = 0` + `AND last_seen < 60 seconds ago` (hide offline / non-active nodes)

File: frontend/src/components/dashboard/DashboardHeader.tsx [hunk 4]
```diff
@@ -107,11 +107,11 @@ export default class DashboardHeader extends React.Component<State> {
             <Link href="/nodes">
               <a>
                 <CardCell
-                  title="Nodes Online"
+                  title="Validators/ Online Nodes"
```
Comment (frol) [inline frontend/src/components/dashboard/DashboardHeader.tsx:110]: Meh, it broke the table since the name is too long. Let's keep "Nodes Online" and later add the (i) helper button (#237) with a popup explaining the numbers.

File: frontend/src/libraries/explorer-wamp/details.ts [hunk 5]
```diff
@@ -2,7 +2,7 @@ import { ExplorerApi } from ".";
 
 export interface Details {
   accountsCount: number;
-  totalNodesCount: number;
+  ValidatorCount: number;
```
Comment (frol) [inline frontend/src/libraries/explorer-wamp/details.ts:5]: JS code style suggests using cameCase starting with the lowercase letter for variable names and fields.

File: frontend/src/libraries/explorer-wamp/nodes.ts [hunk 6]
```diff
@@ -7,36 +7,65 @@ export interface NodeInfo {
...
-                    ${endTimestamp ? `WHERE last_seen < :endTimestamp` : ""}
-                    ORDER BY last_seen DESC
-                    Limit :limit`,
+          last_seen as lastSeen, last_height as lastHeight, last_hash as lastHash,
+          agent_name as agentName, agent_version as agentVersion, agent_build as agentBuild,
+          peer_count as peerCount, is_validator as isValidator, status
+              FROM nodes
+              ${whereClause}
+              ORDER BY last_seen DESC
+              Limit :limit
```
Comment (frol) [inline frontend/src/libraries/explorer-wamp/nodes.ts:42]: `LIMIT`

File: frontend/src/components/nodes/Nodes.tsx [hunk 7]
```diff
@@ -40,35 +105,12 @@ class Nodes extends React.Component<InnerProps> {
...
-            items.map(node => {
-              return <NodeRow key={node.nodeId} node={node} />;
-            })}
-        </FlipMove>
-      </>
+      <FlipMove duration={1000} staggerDurationBy={0}>
+        {items &&
+          items.map(node => {
+            return <NodeRow key={node.nodeId} node={node} />;
+          })}
```
Comment (frol) [inline frontend/src/components/nodes/Nodes.tsx:112]:
```js
          items.map(node => <NodeRow key={node.nodeId} node={node} />)
```

File: frontend/src/components/nodes/Nodes.tsx [hunk 8]
```diff
@@ -10,13 +10,30 @@ import NodeRow from "./NodeRow";
...
 import { OuterProps } from "../accounts/Accounts";
 
-export default class extends React.Component<OuterProps> {
+interface State {
+  validator: boolean;
+}
+export default class extends React.Component<OuterProps, State> {
   static defaultProps = {
-    count: 15
+    count: 3
```
Comment (frol) [inline frontend/src/components/nodes/Nodes.tsx:18]: Why?

File: frontend/src/components/nodes/__test__/NodeRow.test.tsx [hunk 9]
```diff
@@ -12,15 +12,17 @@ describe("<NodeRow />", () => {
...
-            lastSeen: 12345,
-            lastHeight: 12345,
-            lastHash: null,
-            signature: null,
-            agentName: null,
-            agentVersion: null,
-            agentBuild: null,
-            peerCount: null,
-            isValidator: null
+            lastSeen: Date.now(),
```
Comment (frol) [inline frontend/src/components/nodes/__test__/NodeRow.test.tsx:15]: I expect it to make the test flaky (failing randomly). Does it work?

File: docker-compose.yml [hunk 10]
```diff
@@ -12,9 +12,9 @@ services:
   backend:
     build: ./backend
     environment:
-      - NEAR_RPC_URL=https://rpc.devnet.nearprotocol.com
+      # - NEAR_RPC_URL=https://rpc.devnet.nearprotocol.com
       #- NEAR_RPC_URL=https://rpc.betanet.nearprotocol.com
-      #- NEAR_RPC_URL=https://rpc.nearprotocol.com
+      - NEAR_RPC_URL=https://rpc.testnet.nearprotocol.com
```
Comment (frol) [inline docker-compose.yml:17]: Revert these changes.

File: frontend/src/components/nodes/__test__/NodeRow.test.tsx (hunk 9 above)
Comment (icerove) [reply frontend/src/components/nodes/__test__/NodeRow.test.tsx:15]: I never see it flaky...

File: frontend/src/components/nodes/Nodes.tsx (hunk 8 above)
Comment (icerove) [reply frontend/src/components/nodes/Nodes.tsx:18]: use to test and reverted

File: frontend/src/components/nodes/NodeRow.tsx [hunk 11]
```diff
@@ -16,31 +16,45 @@ export default class extends React.PureComponent<Props> {
...
+                  }`}
+                </Col>
+                <Col>
                   <img
-                    src="/static/images/icon-m-user.svg"
+                    src="/static/images/icon-m-block.svg"
                     style={{ width: "12px" }}
                   />
-                  {` @${node.accountId}`}
+                  {`  ${node.lastHeight}`}
```
Comment (frol) [inline frontend/src/components/nodes/NodeRow.tsx:51]: Do we need two spaces before the height?

File: frontend/src/components/nodes/Nodes.tsx [hunk 12]
```diff
@@ -10,13 +10,30 @@ import NodeRow from "./NodeRow";
 
 import { OuterProps } from "../accounts/Accounts";
 
-export default class extends React.Component<OuterProps> {
+interface State {
+  validator: boolean;
```
Comment (frol) [inline frontend/src/components/nodes/Nodes.tsx:14]: Naming nit: Validator is a role. It is odd to see the type of `validator` being `boolean`.

File: frontend/src/components/nodes/Nodes.tsx [hunk 13]
```diff
@@ -28,7 +45,55 @@ export default class extends React.Component<OuterProps> {
...
+            </Button>
+            <Button
+              onClick={this.fetchAllNodes}
+              className="node-selector pagination-total align-self-center"
+            >
+              <img
+                src={"/static/images/icon-m-node-online-gray.svg"}
+                style={{ width: "12px", marginRight: "10px" }}
+              />
+              ALL NODES
```
Comment (frol) [inline frontend/src/components/nodes/Nodes.tsx:70]: https://github.com/nearprotocol/near-explorer/issues/128#issuecomment-617803938

File: frontend/src/components/nodes/Nodes.tsx [hunk 14]
```diff
@@ -28,7 +45,55 @@ export default class extends React.Component<OuterProps> {
   autoRefreshNodes = autoRefreshHandler(Nodes, this.config);
 
   render() {
-    return <this.autoRefreshNodes />;
+    return (
+      <>
+        <div>
+          <Row>
+            <Button
+              onClick={this.fetchValidatorNodes}
```
Comment (frol) [inline frontend/src/components/nodes/Nodes.tsx:53]: You are simulating pages here. Instead of abusing state, let's just have this encoded as the page props, so:

* `/nodes/validators`
* `/nodes/non-validators`

Rename `src/pages/nodes/index.jsx` into `src/pages/nodes/[role].jsx` and handle `role` in `getInitialProps` (you can see how we handle `hash` in `src/pages/blocks/[hash].jsx`), and pass it down to the `Nodes.tsx`.

**Comment by icerove** [comment]
the test for [role].jsx is not passed and I think we can just ignore it as I create a new PR to update prettier

**Comment by frol** [comment]
force your local prettier to format the code, run from the root of the project:

```
$ npm run fmt
```

It should format all the files in the repo.

File: frontend/src/components/nodes/NodeNav.tsx [hunk 15]
```diff
@@ -0,0 +1,108 @@
+import Link from "next/link";
+
+import React from "react";
+import { Row, Col, Spinner } from "react-bootstrap";
+
+import NodesApi from "../../libraries/explorer-wamp/nodes";
+
+interface State {
+  validators?: number;
+  others?: number;
```
Comment (frol) [inline frontend/src/components/nodes/NodeNav.tsx:10]: These are not self-descriptive names.

* `validatorsCount`
* `nonValidatorsCount` is better than `other`

File: frontend/src/components/nodes/NodeNav.tsx [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+            <a style={{ textDecoration: "none", marginLeft: "15px" }}>
+              <Col className="node-selector pagination-total align-self-center">
+                <img
+                  src={"/static/images/icon-m-node-non-validator-status.svg"}
+                  style={{ width: "12px", marginRight: "10px" }}
+                />
+                {others ? (
+                  `${others}  NON-VALIDATOR NODES`
+                ) : (
+                  <Spinner animation="border" variant="secondary" />
```
Comment (frol) [inline frontend/src/components/nodes/NodeNav.tsx:79]: Let's just put a little dash (`-`) when there is no count available yet, there is no reason to overdo with the spinners

File: frontend/src/components/nodes/NodeNav.tsx [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+                )}
+              </Col>
+            </a>
+          </Link>
+          <Link href="/nodes/[role]" as={`/nodes/non-validators`}>
+            <a style={{ textDecoration: "none", marginLeft: "15px" }}>
+              <Col className="node-selector pagination-total align-self-center">
+                <img
+                  src={"/static/images/icon-m-node-non-validator-status.svg"}
```
Comment (frol) [inline frontend/src/components/nodes/NodeNav.tsx:73]: 1. Let's remove the dot indicators from these buttons as we decided to use it for actual online status indication on the Dashboard "Nodes Online".
2. Remove the non-validator-status svg
3. Rename the validator-status svg back to online... :disappointed:

File: frontend/src/components/nodes/NodeNav.tsx [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+                  `${validators}  VALIDATOR NODES`
+                ) : (
+                  <Spinner animation="border" variant="secondary" />
+                )}
+              </Col>
+            </a>
+          </Link>
+          <Link href="/nodes/[role]" as={`/nodes/non-validators`}>
+            <a style={{ textDecoration: "none", marginLeft: "15px" }}>
+              <Col className="node-selector pagination-total align-self-center">
```
Comment (frol) [inline frontend/src/components/nodes/NodeNav.tsx:71]: We should have an indication of which tab is currenlty active. See the provided design

File: frontend/src/components/nodes/Nodes.tsx [hunk 19]
```diff
@@ -10,13 +9,17 @@ import NodeRow from "./NodeRow";
...
 
-export default class extends React.Component<OuterProps> {
+interface Props extends OuterProps {
+  role: string;
+}
+export default class extends React.Component<Props> {
   static defaultProps = {
-    count: 15
+    count: 15,
+    role: "validators"
```
Comment (frol) [inline frontend/src/components/nodes/Nodes.tsx:18]: It does not make sense to have it default.

File: frontend/src/components/nodes/NodeRow.tsx [hunk 20]
```diff
@@ -16,31 +16,45 @@ export default class extends React.PureComponent<Props> {
...
             <Col className="node-row-text">
               <Row>
-                <Col md="auto">#{node.lastHeight}</Col>
-                <Col md="auto" className="pl-0">
+                <Col md={5}>
+                  <img
+                    src="/static/images/icon-m-size.svg"
+                    style={{ width: "12px" }}
+                  />
+                  {`@  ${node.agentName} | ver.  ${node.agentVersion} build  ${
```
Comment (frol) [inline frontend/src/components/nodes/NodeRow.tsx:42]: 1. Let's remove `@  `. It is not an account, so we don't want to confuse people.
2. Remove the extra space after `ver.`

File: frontend/src/libraries/explorer-wamp/details.ts [hunk 21]
```diff
@@ -2,7 +2,7 @@ import { ExplorerApi } from ".";
 
 export interface Details {
   accountsCount: number;
-  totalNodesCount: number;
+  validatorCount: number;
```
Comment (frol) [inline frontend/src/libraries/explorer-wamp/details.ts:5]: Following the style of the above variable name, I would call it with `s` at the end: `validatorsCount`

File: frontend/src/libraries/explorer-wamp/nodes.ts [hunk 22]
```diff
@@ -7,36 +7,78 @@ export interface NodeInfo {
...
       ]);
+      return nodes as NodeInfo[];
     } catch (error) {
       console.error("Nodes.getNodes failed to fetch data due to:");
       console.error(error);
       throw error;
     }
   }
+
+  async getTotalValidatorsAndOther() {
```
Comment (frol) [inline frontend/src/libraries/explorer-wamp/nodes.ts:61]: I would call it `getOnlineNodesStats` and return an object with clearly named elements instead of an array.

File: frontend/src/libraries/explorer-wamp/nodes.ts [hunk 23]
```diff
@@ -7,36 +7,78 @@ export interface NodeInfo {
...
+    try {
+      const [validators, others] = await Promise.all([
+        this.call("select", [
+          `SELECT COUNT(*) as total FROM nodes
+              WHERE is_validator = 1
+          `
+        ]).then((it: any) => it[0].total),
+        this.call("select", [
+          `SELECT COUNT(*) as total FROM nodes
+              WHERE is_validator = 0
```
Comment (frol) [inline frontend/src/libraries/explorer-wamp/nodes.ts:71]: You can put these into a single SQL query as it is done in the `details.js`

File: frontend/src/pages/nodes/[role].jsx [hunk 24]
```diff
@@ -1,17 +1,29 @@
...
+
 import Nodes from "../../components/nodes/Nodes";
 import Content from "../../components/utils/Content";
 
 export default class extends React.Component {
+  static async getInitialProps({ query: { role } }) {
+    try {
+      return { role }
+    } catch (err) {
+      return { role, err }
```
Comment (frol) [inline frontend/src/pages/nodes/[role].jsx:15]: When do you expect `return { role }` to fail?

File: frontend/src/pages/nodes/[role].jsx [hunk 25]
```diff
@@ -1,17 +1,29 @@
 import Head from "next/head";
 
+import React from "react";
+
+import NodeNav from "../../components/nodes/NodeNav"
```
Comment (frol) [inline frontend/src/pages/nodes/[role].jsx:5]: This import belongs to the group below.

File: frontend/src/components/nodes/NodeNav.tsx (hunk 18 above)
Comment (icerove) [reply frontend/src/components/nodes/NodeNav.tsx:67]: they all active, one is active validators, another one is active non validators.

File: frontend/src/components/nodes/NodeNav.tsx [hunk 26]
```diff
@@ -58,11 +58,7 @@ export default class extends React.Component<State> {
                   src={"/static/images/icon-m-node-validator-status.svg"}
                   style={{ width: "12px", marginRight: "10px" }}
                 />
-                {validators ? (
-                  `${validators}  VALIDATOR NODES`
-                ) : (
-                  <Spinner animation="border" variant="secondary" />
-                )}
+                {validatorsCount ? `${validatorsCount}  VALIDATOR NODES` : `-`}
```
Comment (frol) [inline frontend/src/components/nodes/NodeNav.tsx:61]: I meant `- VALIDATOR NODES`.

Also, given our today's sync, let's call the tabs as "Validating" | "Non-Validating"

File: frontend/src/components/nodes/NodeNav.tsx (hunk 18 above)
Comment (icerove) [reply frontend/src/components/nodes/NodeNav.tsx:67]: I change it a little bit since if there if border, it is easy to see where to click and also good for mobile. I will change border color when actived.

File: frontend/src/components/nodes/NodeNav.tsx (hunk 17 above)
Comment (icerove) [reply frontend/src/components/nodes/NodeNav.tsx:69]: it will be weird when there is no icon, so I change the icon instead.

Comment (frol) [reply frontend/src/components/nodes/NodeNav.tsx:69]: I see, good point. Let's use the green circle then indicating the online status and given we filter out offline nodes, we only need the green circle.

**Comment by frol** [comment]
@icerove Bug report: When I click on the button, it changes the border color to be blue, but once I click anywhere else, the "selected" button gets the gray border back:

image

Let's use the green online circle on the buttons instead of the block icon.

File: frontend/src/components/nodes/NodeNav.tsx [hunk 27]
```diff
@@ -0,0 +1,136 @@
...
+  changeNonValidatorNodeBorder = () => {
+    const validatorDiv = document.getElementById("validator-node");
+    if (validatorDiv) {
+      validatorDiv.style.border = "2px solid #e6e6e6";
+    }
+    const nonValidatorDiv = document.getElementById("non-validator-node");
+    if (nonValidatorDiv) {
+      nonValidatorDiv.style.border = "2px solid #0066ff";
+    }
+  };
```
Comment (frol) [inline frontend/src/components/nodes/NodeNav.tsx:66]: It can be done muuuuch simpler by just settings `selected` CSS class on the relevant button based on the `props.role === 'validating'`

Comment (icerove) [reply frontend/src/components/nodes/NodeNav.tsx:66]: okay

**Comment by jakestutzman** [comment]
Need to fix this spacing. These tabs need bottom margin. Also, I wouldn't put green circles on both tabs. It doesn't make sense.
Screen Shot 2020-04-28 at 6 27 01 AM

**Comment by frol** [comment]
@jakestutzman Thanks! I have created an issue to track it https://github.com/nearprotocol/near-explorer/issues/270
